### PR TITLE
[Feat] Better workflow for adding contacts from emails

### PR DESCRIPTION
**Pull request by henochit**
## Pullrequest
When I click on an email, offers to user to add to contacts.

If a contact already exists with that email, it should offer a link (or perhaps many) to review the rest of the contact information.
If no contact exists with that email, it should offer me to add it as a contact.

related to:

https://avan.tech/item40524

### Issues

- [X] None

### Checklist

- [X] None

### How2Test

- [X] Open an email
- [X] Click the sender. If the sender is not a local contact, they should offer a button to add the sender as a contact. If the contact exists, you will see the rest of the coordinates.

### Todo

- [X] None

### CODE REVIEW

File: modules/contacts/hm-contacts.php [hunk 1]
```diff
@@ -25,11 +25,24 @@ public function add_contact($data) {
         return true;
     }
 
-    public function get($id, $default=false) {
-        if (!array_key_exists($id, $this->contacts)) {
-            return $default;
+    public function get($id, $email_address="", $default=false) {
```
Comment (kroky) [inline modules/contacts/hm-contacts.php:28]: It is not advisable to change signature of a function/method (reorder parameters) unless you made sure you updated all places that call this method.

File: modules/contacts/hm-contacts.php [hunk 2]
```diff
@@ -25,11 +25,24 @@ public function add_contact($data) {
...
+        if(array_key_exists($id, $this->contacts)) {
```
Comment (kroky) [inline modules/contacts/hm-contacts.php:29]: There are various spacing issues with the code, please run through a suitable editor or manually address. Try to be in sync with existing code - e.g. space after if/for/etc. operators, no space after function and before () brackets. See the diff here on github and you will notice all these places.

File: modules/imap/site.js [hunk 3]
```diff
@@ -1,1102 +1,1320 @@
-'use strict';
+"use strict";
```
Comment (kroky) [inline modules/imap/site.js:1]: There are quite many spacing changes here that are not related to the PR - please reset and only do the changes you want. E.g. changing from 4 to 2 spaces for the whole file is not desirable.

File: scripts/hm3.ini [hunk 4]
```diff
@@ -0,0 +1,900 @@
+
```
Comment (kroky) [inline scripts/hm3.ini:1]: This is .gitignore-d file and shouldn't be in the repo.

File: scripts/hm3.rc [hunk 5]
```diff
@@ -0,0 +1 @@
...tus","ajax_pop3_debug","ajax_pop3_message_display","ajax_pop3_folder_display","ajax_pop3_unread","ajax_pop3_status","ajax_imap_debug","ajax_imap_status","ajax_imap_unread","ajax_imap_sent","ajax_imap_flagged","ajax_imap_folder_expand","ajax_imap_folder_display","ajax_imap_combined_inbox","ajax_imap_search","ajax_unread_count","ajax_imap_message_content","ajax_imap_save_folder_state","ajax_imap_message_action","ajax_imap_delete_message","ajax_imap_archive_message","ajax_imap_flag_message","ajax_imap_update_combined_source","ajax_imap_mark_as_read","ajax_imap_move_copy_action","ajax_imap_folder_status","ajax_clear_attachment_chunks","ajax_smtp_debug","ajax_smtp_save_draft","ajax_smtp_delete_draft","ajax_profiles_status","ajax_get_test_chunk","ajax_upload_chunk","accounts","change_password","ajax_no_op","calendar","ajax_nux_service_select","ajax_nux_add_service","history","ajax_save_search","ajax_update_search","ajax_delete_search","ajax_update_save_search_label","advanced_search","ajax_adv_search","highlights","profiles","folders","ajax_imap_folders_delete","ajax_imap_folders_create","ajax_imap_folders_rename","ajax_imap_special_folder","ajax_imap_clear_special_folder","shortcuts"]}}
```
Comment (kroky) [inline scripts/hm3.rc:1]: This is .gitignore-d file and shouldn't be in the repo.

File: modules/contacts/hm-contacts.php (hunk 2 above)
Comment (henochit) [reply modules/contacts/hm-contacts.php:29]: Done

**Comment by kroky** [comment]
@henochit , please let me know when this is ready for review.

**Comment by henochit** [comment]
@kroky ready to be reviewed.

File: modules/contacts/site.js [hunk 6]
```diff
@@ -20,12 +20,57 @@ var delete_contact = function(id, source, type) {
...
+    if (contact) {
+      Hm_Ajax.request(
+        [
+          { name: 'hm_ajax_hook', value: 'ajax_add_contact' },
+          { name: 'contact_value', value: contact },
+          { name: 'contact_source', value: source },
+        ],
+        function (res) {
+          $('.add_contact_controls').toggle();
+          window.location.reload();
```
Comment (kroky) [inline modules/contacts/site.js:33]: If you reload the whole page, the other JS operations you do seems unneeded. Why do we need to reload the whole page? Can't we do the updates here with JS?

File: modules/contacts/hm-contacts.php (hunk 2 above)
Comment (kroky) [reply modules/contacts/hm-contacts.php:29]: @henochit now it is all good but the imap/site.js file - it went from 4 to 2 spaces for indentation and also some single quotes changed to double. Can you please reset the file to its original content and apply only the changes you want to apply and not such global formatting changes?

File: modules/imap/site.js (hunk 3 above)
Comment (kroky) [reply modules/imap/site.js]: This file is not yet resolved - see diff section of this PR.

**Comment by kroky** [comment]
@henochit please also rebase master as there are conflicts to resolve in imap/site.css

File: modules/contacts/site.js [hunk 7]
```diff
@@ -20,12 +20,57 @@ var delete_contact = function(id, source, type) {
...
+        function (res) {
+          $('.popup .show').removeClass('show');
+          window.location.reload();
+          remove_message_content();
+        }
+      );
+    }
+  };
+
+var add_contact_from_popup = function() {
```
[the remark this answers is not in the document]
Comment (henochit) [reply modules/contacts/site.js:60]: it could be from a bad fusion, let me check.

File: modules/contacts/site.js (hunk 6 above)
Comment (henochit) [reply modules/contacts/site.js:33]: I reload to refresh the session storage, I follow the way they implemented `add_contact_from_message_view`. is there a way to refresh the cache page from session storage without reloading the page?

Comment (kroky) [reply modules/contacts/site.js:33]: Session storage is available at all times - you don't need to reload the page to get the latest data in session storage. If you need to refresh specific sections on the page, you can use JS functions for this or ajax requests.

File: .gitignore [hunk 8]
```diff
@@ -23,3 +23,5 @@ scripts/test.php
 *.sublime-workspace
 composer.phar
 lib/hm3/users/
+/scripts/hm3.ini
+/scripts/hm3.rc
```
Comment (kroky) [inline .gitignore:27]: What are these? We don't store hm3 files in scripts folder.

**Comment by henochit** [comment]
@kroky, ready to be merged.
